src/Agent.Worker/Build/FileContainerServer.cs

```diff
@@ -329,55 +330,70 @@ private async Task<UploadResult> BlobUploadAsync(IAsyncCommandContext context, I
-            var associateMonitor = AssociateReportingAsync(context, files.Count(), uploadFinished, token);
+                // Start associate monitor
+                var uploadFinished = new TaskCompletionSource<int>();
+                var associateMonitor = AssociateReportingAsync(context, files.Count(), uploadFinished, token);
 
-            // Start parallel associate tasks.
-            var parallelAssociateTasks = new List<Task<UploadResult>>();
-            for (int uploader = 0; uploader < concurrentUploads; uploader++)
-            {
-                parallelAssociateTasks.Add(AssociateAsync(context, queue, token));
-            }
+                // Start parallel associate tasks.
+                var parallelAssociateTasks = new List<Task<UploadResult>>();
+                for (int uploader = 0; uploader < concurrentUploads; uploader++)
+                {
+                    parallelAssociateTasks.Add(AssociateAsync(context, queue, token));
+                }
 
-            // Wait for parallel associate tasks to finish.
-            await Task.WhenAll(parallelAssociateTasks);
-            foreach (var associateTask in parallelAssociateTasks)
-            {
-                // record all failed files.
-                uploadResult.AddUploadResult(await associateTask);
-            }
+                // Wait for parallel associate tasks to finish.
+                await Task.WhenAll(parallelAssociateTasks);
+                foreach (var associateTask in parallelAssociateTasks)
+                {
+                    // record all failed files.
+                    uploadResult.AddUploadResult(await associateTask);
+                }
 
-            // Stop monitor task
-            uploadFinished.SetResult(0);
-            await associateMonitor;
+                // Stop monitor task
+                uploadFinished.SetResult(0);
+                await associateMonitor;
 
-            // report telemetry
-            if (!Guid.TryParse(context.GetVariableValueOrDefault(WellKnownDistributedTaskVariables.PlanId), out var planId))
-            {
-                planId = Guid.Empty;
+                // report telemetry
+                if (!Guid.TryParse(context.GetVariableValueOrDefault(WellKnownDistributedTaskVariables.PlanId), out var planId))
+                {
+                    planId = Guid.Empty;
+                }
+                if (!Guid.TryParse(context.GetVariableValueOrDefault(WellKnownDistributedTaskVariables.JobId), out var jobId))
+                {
+                    jobId = Guid.Empty;
+                }
+                await clientTelemetry.CommitTelemetryUpload(planId, jobId);
             }
-            if (!Guid.TryParse(context.GetVariableValueOrDefault(WellKnownDistributedTaskVariables.JobId), out var jobId))
+            catch
             {
-                jobId = Guid.Empty;
+                var blobStoreHost = dedupClient.Client.BaseAddress.Host;
```

Comment:
This block of code looks similar to the one in FileContainerProvider - could it make sense to move it to the separate method?

Reply:
Yes, it is similar and I've thought about implementing of common class for generating warnings for BlobStore. But the problem with creating any common method for localization strings generating is that we have the [test](https://github.com/microsoft/azure-pipelines-agent/blob/b93f8f1f4e53ddf5d9cd25090db6f71b092d177c/src/Test/L0/LocStringsL0.cs#L72) that checks for the unused localization strings, and the logic of the test doesn't presuppose that variables would be passed into Loc method, it checks only direct passing of localization aliases, without variables.